### PR TITLE
Bundle rework

### PR DESCRIPTION
**Pull request by rm-dr**
Our current bundle scheme is causing quite a few bugs. This PR (paired with tectonic-typesetting/tectonic-texlive-bundles#14) resolves many of these by implementing proper search paths in bundles.

These changes are **not** be breaking. Old zip and bundles will work with this PR.
(I have, however, removed the scripts that create them. They should eventually be deprecated.)

Changes in this PR:
 - A new bundle format (see tectonic-typesetting/tectonic-texlive-bundles#14)
 - File search paths (also documented in the bundle repo)
 - Bundle autodetection from url (thus, `--web-bundle` has been removed, and `--bundle` now handles both URLs and paths)
 - Bundle caching rework (mostly invisible to the user, enables cleaner code)
 - Cache organization rework (the bundle cache needs fewer files and is in its own subdir)
 - Many minor bundle-related cleanups and optimizations

Most changes in this PR are in the `bundles` crate. Changes in other files are usually minor side-effects thanks to a slightly different API.

### CODE REVIEW

**Review by vlasakm — COMMENTED** [review]
Just some things I noticed at firts glance.

File: crates/bundles/src/zip.rs [hunk 1]
```diff
@@ -18,19 +22,155 @@ use crate::Bundle;
...
+        };
+
+        // If we don't have this path in the index, this file doesn't exist.
+        // Make sure the bundle script adds ALL files to the index!
+        //
+        // The code below takes care to clone these strings only if it needs to.
+        let paths: &Vec<String> = match self.index.get(n) {
+            Some(s) => s,
+            None => return None,
+        };
```
[the remark this answers is not in the document]
Comment (rm-dr) [reply crates/bundles/src/zip.rs]: Oops. Thank you!

**Comment by pkgw** [comment]
Thanks for all this work! As I remarked in the `texlive-bundles` PR, I wasn't getting any notifications about your work on this topic. I hope to start reading over your PRs and understanding what you've been working on, although to be honest it will likely take me a lot longer to do so than I'd like.

**Comment by rm-dr** [comment]
Alright, I think I've implemented all the extra bundle features we need to make Tectonic's bundles actually work.

The code here is very close to done, except for a few minor problems (listed at the top) and whatever edits we come up with.

File: crates/bundles/src/itar.rs [hunk 2]
```diff
@@ -3,198 +3,293 @@
...
-/// A simple web-based file backend based on HTTP Range requests.
-///
-/// This type implements the [`CacheBackend`] trait and so can be used for
-/// web-based bundle access thorugh the [`crate::cache::CachingBundle`]
-/// framework.
-#[derive(Debug)]
-pub struct IndexedTarBackend {
-    reader: DefaultRangeReader,
+impl ItarFileIndex {
+    fn new() -> Self {
```
Comment (pkgw) [inline crates/bundles/src/itar.rs:55]: An argument-less `new()` method is a signpost that you should be implementing the `Default` trait; I think you can auto-derive Default for this type and `ItarFileIndex`.

**Comment by pkgw** [comment]
Thanks again for your work on this.

Will this new caching scheme work if multiple Tectonic processes are using and writing to the cache simultaneously? This scenario can easily happen if you have a work tree with several documents asnd Tectonic wired into a Make or Ninja build tool that runs multiple jobs at the same time.

At the moment, it looks like this PR is going to need updating to fix conflicts with #1132 . It will also need a pass to fix a bunch of Clippy complaints, and the test failure(s) of course. At the moment I see two tests failing on my laptop: `test_cache_location_redirect` and `test_bundle_update`. Unfortunately the Azure build logs have expired (and we can't get new build logs until the merge conflicts are resolved), so I don't know if that aligns with what the CI was finding.

**Comment by rm-dr** [comment]
> Will this new caching scheme work if multiple Tectonic processes are using and writing to the cache simultaneously?

No clue. I'll look into this later.
I'm not sure the previous cache was fully safe... It should be ok most of the time, since cache files are only read, but two processes downloading the same file at the same time could cause problems.


As for cleanup, I'll get to that eventually. Cleanup is the reason tectonic-typesetting/tectonic-texlive-bundles#14 is still in draft mode. This PR and that PR solve the same problem.

**Comment by rm-dr** [comment]
Alright, that should be it. The new cache is multi-process-safe via renames, tests pass, and clippy errors have been fixed. Both this and tectonic-typesetting/tectonic-texlive-bundles#14 should be ready to merge.

**Comment by pkgw** [comment]
CC @CraftSpider here too.

**Comment by rm-dr** [comment]
Hmmm
A few tests are still failing, not sure why.

**Comment by pkgw** [comment]
Hmm, I'm not sure why this error would only be showing up in the cross-tests. The one thing that I can think of is that those run through the `cross` program, which filters the environment variables that propagate down to the subprograms that it runs ... but I don't see anything about this particular error that would depend on environment variables being set.

**Comment by rm-dr** [comment]
Just fixed #74. File names with spaces are now safe in ttbv1 bundles---but newlines will still cause errors.

I'm not sure that's a problem, though, since I've never seen a legitimate use case for files with newlines.

**Comment by rm-dr** [comment]
Here's an idea: should this new bundle format also include latex format files?

Now that the bundle-building code is in Rust, it shouldn't be too hard to generate format files while we build the bundle. (assuming format files are identical on all machines)


Would be nice to do that here, so that we only have one round of big, possibly disruptive changes.

**Pull request closed without merging**